base_layer/wallet/src/transaction_manager.rs

```diff
@@ -0,0 +1,529 @@
+}
+
+impl TransactionManager {
+    pub fn new() -> TransactionManager {
+        TransactionManager {
+            pending_outbound_transactions: HashMap::new(),
+            pending_inbound_transactions: HashMap::new(),
+            completed_transactions: HashMap::new(),
+        }
+    }
+
+    /// Start to send a new transaction.
+    /// Accepts a well formed SenderTransactionProtocol.
+    /// Returns the public SenderMessage to be transmitted to the recipient.
+    pub fn start_send_transaction(
+        &mut self,
+        mut sender_transaction_protocol: SenderTransactionProtocol,
+    ) -> Result<SenderMessage, TransactionManagerError>
+    {
+        if !sender_transaction_protocol.is_single_round_message_ready() {
+            return Err(TransactionManagerError::InvalidStateError);
+        }
+
+        let msg = sender_transaction_protocol.build_single_round_message()?;
+
+        self.pending_outbound_transactions
+            .insert(msg.tx_id.clone(), sender_transaction_protocol);
+
+        Ok(SenderMessage::Single(Box::new(msg)))
+    }
+
+    /// Accept the public reply from a recipient and apply the reply to the relevant transaction protocol
+    pub fn accept_recipient_reply(
+        &mut self,
+        recipient_reply: RecipientSignedTransactionData,
+    ) -> Result<(), TransactionManagerError>
+    {
+        let mut marked_for_removal = None;
+
+        for (tx_id, stp) in self.pending_outbound_transactions.iter_mut() {
+            if stp.check_tx_id(recipient_reply.tx_id) && stp.is_collecting_single_signature() {
+                stp.add_single_recipient_info(recipient_reply.clone())?;
+                stp.finalize(KernelFeatures::empty())?;
+                let tx = stp.get_transaction()?;
+                self.completed_transactions.insert(recipient_reply.tx_id, tx.clone());
+                marked_for_removal = Some(tx_id.clone());
+                break;
+            }
+        }
+
+        if marked_for_removal.is_none() {
+            return Err(TransactionManagerError::TransactionDoesNotExistError);
+        }
+
+        if let Some(tx_id) = marked_for_removal {
+            self.pending_outbound_transactions.remove(&tx_id);
+        }
+
+        Ok(())
+    }
+
+    /// Accept a new transaction from a sender by handling a public SenderMessage
+    /// Returns the public reply message to be sent back to the sender.
+    pub fn accept_transaction(
+        &mut self,
+        sender_message: SenderMessage,
+        nonce: SecretKey,
+        spending_key: SecretKey,
+    ) -> Result<RecipientSignedTransactionData, TransactionManagerError>
+    {
+        let rtp = ReceiverTransactionProtocol::new(sender_message, nonce, spending_key, OutputFeatures::empty());
+        let recipient_reply = rtp.get_signed_data()?;
+
+        // Check this is not a repeat message i.e. tx_id doesn't already exist in our pending or completed transactions
+        for (tx_id, _) in self.pending_outbound_transactions.iter() {
```

Comment:
```suggestion
        if self.pending_outbound_transactions.contains_key(&tx_id) {
```
will be at least as fast as this O(n) search, possibly faster (depends how the hashmap stores keys)

Reply:
Correct